File: app/src/main/java/com/starsep/sokoban/gamelogic/Tile.java
package com.starsep.sokoban.gamelogic;

import android.util.Log;

import com.starsep.sokoban.Sokoban;

public class Tile {
	private final static int GRASS_MASK = 0;
	public final static int WALKABLE_MASK = 0;
	private final static int SOLID_MASK = 1;
	private final static int GROUND_MASK = 2;
	private final static int ENDPOINT_MASK = 4;
	private final static int CRATE_MASK = 8;

	public final static char wall = 'W';
	public final static char crate = '#';
	public final static char ground = '.';
	public final static char endpoint = 'X';
	public final static char grass = '_';
	public final static char crateOnEndpoint = '&';

	public static boolean isWalkable(char c) {
		return (mask(c) & SOLID_MASK) == WALKABLE_MASK;
	}

	public static boolean isCrate(char c) {
		return (mask(c) & CRATE_MASK) == CRATE_MASK;
	}

	public static int mask(char c) {
		switch (c) {
			case wall:
				return SOLID_MASK;
			case crate:
				return GROUND_MASK | CRATE_MASK | SOLID_MASK;
			case ground:
				return GROUND_MASK;
			case endpoint:
				return GROUND_MASK | ENDPOINT_MASK;
			case grass:
				return GRASS_MASK;
			case crateOnEndpoint:
				return mask(endpoint) | mask(crate);
			default: {
				Log.e(Sokoban.TAG, "Tile.mask: " + "Unknown tile " + c);
				return WALKABLE_MASK;
			}
		}
	}

	public static char maskToChar(int m) {
		char[] tiles = {wall, crate, ground, endpoint, grass, crateOnEndpoint};
		for (char c : tiles) {
			if (m == mask(c)) {
				return c;
			}
		}
		Log.e(Sokoban.TAG, "Tile.maskToChar: " + "Unknown mask " + m);
		return grass;
	}

	public static boolean isGrass(char c) {
		return c == grass;
	}
	private static int crateMask() {
		return Tile.CRATE_MASK | Tile.SOLID_MASK;
	}

	public static char withoutCrate(char tile) {
		return maskToChar(mask(tile) ^ crateMask());
	}

	public static char withCrate(char tile) {
		return maskToChar(mask(tile) | crateMask());
	}



}

File: app/src/main/kotlin/com.starsep.sokoban.release/gamelogic/Tile.kt
package com.starsep.sokoban.release.gamelogic

import android.util.Log

import com.starsep.sokoban.release.Sokoban

object Tile {
    val WALKABLE_MASK = 0
    val wall = 'W'
    val crate = '#'
    val ground = '.'
    val endpoint = 'X'
    val grass = '_'
    val crateOnEndpoint = '&'
    private val GRASS_MASK = 0
    private val SOLID_MASK = 1
    private val GROUND_MASK = 2
    private val ENDPOINT_MASK = 4
    private val CRATE_MASK = 8

    fun isWalkable(c: Char): Boolean {
        return mask(c) and SOLID_MASK == WALKABLE_MASK
    }

    fun isCrate(c: Char): Boolean {
        return mask(c) and CRATE_MASK == CRATE_MASK
    }

    fun mask(c: Char): Int {
        return when (c) {
            wall -> SOLID_MASK
            crate -> GROUND_MASK or CRATE_MASK or SOLID_MASK
            ground -> GROUND_MASK
            endpoint -> GROUND_MASK or ENDPOINT_MASK
            grass -> GRASS_MASK
            crateOnEndpoint -> mask(endpoint) or mask(crate)
            else -> {
                Log.e(Sokoban.TAG, "Tile.mask: " + "Unknown tile " + c)
                WALKABLE_MASK
            }
        }
    }

    fun maskToChar(m: Int): Char {
        val tiles = charArrayOf(wall, crate, ground, endpoint, grass, crateOnEndpoint)
        tiles
                .filter { m == mask(it) }
                .forEach { return it }
        Log.e(Sokoban.TAG, "Tile.maskToChar: " + "Unknown mask " + m)
        return grass
    }

    fun isGrass(c: Char): Boolean {
        return c == grass
    }

    private fun crateMask(): Int {
        return Tile.CRATE_MASK or Tile.SOLID_MASK
    }

    fun withoutCrate(tile: Char): Char {
        return maskToChar(mask(tile) xor crateMask())
    }

    fun withCrate(tile: Char): Char {
        return maskToChar(mask(tile) or crateMask())
    }


}
